src/Compilers/CSharp/Portable/BoundTree/Expression.cs

```diff
@@ -0,0 +1,1791 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis.Semantics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using System;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    partial class BoundExpression : IExpression
+    {
+        ITypeSymbol IExpression.ResultType
+        {
+            get { return this.Type; }
+        }
+
+        OperationKind IOperation.Kind
+        {
+            get { return this.ExpressionKind; }
+        }
+
+        object IExpression.ConstantValue
+        {
+            get
+            {
+                ConstantValue value = this.ConstantValue;
+                if (value == null)
+                {
+                    return null;
+                }
+
+                return value.Value;
+            }
+        }
+
+        SyntaxNode IOperation.Syntax
+        {
+            get { return this.Syntax; }
+        }
+
+        protected virtual OperationKind ExpressionKind { get { return OperationKind.None; } }
+        // protected abstract OperationKind ExpressionKind { get; }
+    }
+
+    internal abstract partial class BoundNode : IOperationSearchable
+    {
+        IEnumerable<IOperation> IOperationSearchable.Descendants()
```

Comment:
ImmutableArray may require construction and/or storage of the array instance whereas the IEnumerable only requires yielding values and may map better to the bound node info.   If that's not true, and there is already an underlying ImmutableArray for both languages for all bound node types, then by all means expose ImmutableArray.